### ISSUE TITLE
B3 headers warning with JMS

### ISSUE THREAD

**Issue by saturnism** [spring-cloud/spring-cloud-sleuth#1643]
**Describe the bug**
When using Spring Cloud Sleuth + Spring Integration + JMS + IBM MQ, the B3 headers are propagated to the JMS headers. However, however MQ's `JMSMessage` doesn't allow these headers and throws exception, which generates the warning:

```
WARN [,5ec0b3c28f0ecf2501557752f69a8f32,0e2e4e1c88a37d95,true] 11695 --- [sub-subscriber1] o.s.i.jms.DefaultJmsHeaderMapper         : failed to map Message header 'X-B3-TraceId' to JMS property

com.ibm.msg.client.jms.DetailedMessageFormatException: JMSCC0049: The property name 'X-B3-TraceId' is not a valid Java(tm) identifier.
...
```

I read that the warning can be ignored, but need to be explicitly suppressed from the logs. If suppressing the warning level log from this class, then I might miss other warning messages.

I'm using `IntegrationFlow` with `Jms.outboundGateway`.  In the flow, I tried to use `f.headerFilter` to remove the headers, but that didn't seem to work. I suspect the headers were re-generated in the subsequent `.handle(Jms.outboundGateway()...)`

```
IntegrationFlows.from("output")
  .transform(...)
  .headerFilter(...) // remove B3 headers
  .handle(Jms.outboundGateway()...) // this still gets the header
```

To work around this, I created a new `JmsHeaderMapper` to explicitly remove the B3 header.  The logic is mostly from the `private MessageHeaders(MessageHeaders original, Set<String> keysToIgnore)` constructor, but it's private.

```
class NoB3JmsHeaderMapper extends DefaultJmsHeaderMapper {
    @Override
    public void fromHeaders(MessageHeaders headers, Message jmsMessage) {
      List<String> ignore = Arrays
          .asList("X-B3-TraceId", "X-B3-ParentSpanId", "X-B3-SpanId", "X-B3-Sampled");
      Map<String, Object> filtered = new HashMap<>(headers.size());
      headers.forEach((key, value) -> {
        if (!ignore.contains(key)) {
          filtered.put(key, value);
        }
      });
      super.fromHeaders(new MessageHeaders(filtered), jmsMessage);
    }
  }
```

Checking in to see if there are easier ways I missed.

Thanks!

**Comment by adriancole** [comment]
yeah spring integration etc should have never used hyphenated headers. FWIW in 3.0 the default is corrected to use b3 single when we can't determine if something is a remote span or not

https://github.com/spring-cloud/spring-cloud-sleuth/blob/master/spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/autoconfig/TraceBaggageConfiguration.java#L103-L117

please upvote this as I know some were concerned about instrumentation out there still not understanding b3 single (which is valid, just not sure which.. certainly not brave for recent years) https://github.com/openzipkin/brave/issues/1211

**Comment by adriancole** [comment]
the bigger problem is we can't tell when something will be remote vs an in-memory channel in a lot of the spring abstractions. A couple years ago I started on this, but I didn't follow up. If there's a place you know (or heck can even help!) it must be remote, then we should already b3 using b3 single. For example, the default for JMS should be b3 single https://github.com/openzipkin/brave/pull/572

So.. I suspect your workaround is safe if it only applies out of process, and at least here, it should also apply out of process :)

In fact, probably this part of the change we can do in 2.2.3? wdyt? https://github.com/spring-cloud/spring-cloud-sleuth/pull/1607/files#diff-db43b7e91bd69d063333c20b947c902bR83-R84

**Comment by adriancole** [comment]
PS your workaround, if converted to b3 single would be even better I guess.. but I think in spring-messaging we already have a few types of header formats in use..

**Comment by marcingrzejszczak** [comment]
I'm all for making the change there in 2.2.x (the one you've placed in 3.0) @adriancole

**Comment by adriancole** [comment]
https://github.com/spring-cloud/spring-cloud-sleuth/pull/1655 will fix this in 2.2.x

**Comment by saturnism** [comment]
Awesome! Thanks 😄
